### PR TITLE
refactor: split out core functionality of _trustedboot into a separate _usi feature

### LINKED ISSUE #2443: Allow in-place updates without TPM
- [x] Document new design, where trustedboot is optional: https://github.com/gardenlinux/gardenlinux/blob/feat/usi/docs/boot_modes.md
- [x] Split out core functionality of `_trustedboot` into a separate `_usi` feature: https://github.com/gardenlinux/gardenlinux/commit/98d39efeb42ab2e5a86892de98ec9c600c2d9223
- [x] Implement a `_nocrypt` storage provider for `_usi`: https://github.com/gardenlinux/gardenlinux/commit/a412f07d42ed296e4eb2001ff5c2dbf543bc0c5b
- [x] Refactor old `_boot` feature and related things into `_legacy` feature: https://github.com/gardenlinux/gardenlinux/commit/b0a7dc161170367cb5bb5caaa162981d2804dc63
- [x] ~Make `_usi` the default unless `_legacy` is explicitly set: https://github.com/gardenlinux/gardenlinux/commit/b015521cff9a3d62064e7022c6e6fcf4a0fa64f0~  
  Decided against it, reverted by https://github.com/gardenlinux/gardenlinux/commit/58c1dc83a271ef6e34277a3f275a920d448fb086
- [x] disable secureboot signing when the `_nocrypt` feature is used (or maybe by default unless `_trustedboot` requested): https://github.com/gardenlinux/gardenlinux/commit/aa0d04b6b744ccc933b2b4e14884d6e0104f3d37
- [x] ~Proper security design on how to prevent cross flavour exploits (i.e. `_nocrypt` mode gaining access to `_tpm2` mode keys) (WIP idea: measure the full feature set into PCR15 and bind TPM key to it. Problem: What about timing attacks, blocking TPM write during boot of `_nocrypt` flavour and only releasing after attacker controlled code takes over)~
  Not needed when secureboot signing only done for `_trustedboot` flavors.

### PR DESCRIPTION
**Pull request by nkraetzschmar**
**What this PR does / why we need it**:

Split out the core UKI / USI based boot logic from `_trustedboot` into its own feature: `_usi`.
USI stands for Unified System Image, see the spec linked below for details.

On its own this does not change anything for now, but it will allow for in-place updates down the line without those requiring trustedboot or a TPM device.

As part of this `_boot` was renamed to `_legacy` to distinguish the two boot modes.

See [docs/boot_mode.md](https://github.com/gardenlinux/gardenlinux/blob/58c1dc83a271ef6e34277a3f275a920d448fb086/docs/boot_modes.md) for the design spec.

**Which issue(s) this PR fixes**:
Resolves #2443

**Definition of Done:**
- [x] The code is sufficiently documented
- [x] Shared the changes with the Team so everyone is aware
- [x] The code is appropriately tested
- [x] Checked if the code needs to be backportet to release branches of maintained versions (perform the actual backport after the merge to `main`) -> no backport needed

### CODE REVIEW

File: docs/boot_modes.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+# example disk layout for legacy mode
+
+vda                                                                         
+├─vda1
+│    vfat   FAT32 EFI   C1FE-FA19                                28M    67% /efi
+└─vda2
+     ext4   1.0   ROOT  aea6d56e-a2bf-944c-78fe-184a4d6add9f   29.1G     2% /
+```
+
+## Unified System Image (USI)
```
Comment (fwilhe) [inline docs/boot_modes.md:42]: so usi is a term that we invented, right?

File: docs/boot_modes.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+The design is loosely inspired by Lennart Poettering's ideas[^1] on how modern operating systems should boot.
+We also fully embrace Unified Kernel Images (UKIs)[^2] and we also make use of an immutable root filesystem.
+However, the design differs in how the root filesystem is verified.
+
+In Lennart Poettering's design, the root (or just the `/usr` partition) is stored on a dm-verity disk; the root hash of the dm-verity Merkle tree is then embedded into the UKI.
+While this design works well, it does add quite a bit of complexity.
+It requires at least three partitions (more if you want in-place updates with A/B partitions): EFI system partition, root dm-verity data partition, and root dm-verity hash tree partition.
+
+Due to the incredibly small footprint of our root partition, we can afford to go a different route: we can embed the entire root filesystem into the UKI.
```
Comment (fwilhe) [inline docs/boot_modes.md:52]: > incredibly small footprint

maybe give some numbers here. the limit is the 4gi limit of the FAT on EFI, correct?

File: docs/boot_modes.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+| UEFI boot                | :white_check_mark: | :white_check_mark: |
+| BIOS boot                | :x:                | :white_check_mark: |
+| secureboot / trustedboot | :heavy_check_mark: | :x:                |
+| in-place updates         | :white_check_mark: | :x:                |
+| TPM 2.0 disk encryption  | :heavy_check_mark: | :x:                |
+
+*( :heavy_check_mark: = optional )*
+
+> [!TIP]
+> If you are unsure which mode to pick: go for the *trustedboot* images, these are USIs with all optional security features enabled.
```
Comment (fwilhe) [inline docs/boot_modes.md:22]: who does this address? is this specific to gardener? what about users who build custom images based on Garden Linux?

File: docs/boot_modes.md (hunk 1 above)
Comment (nkraetzschmar) [reply docs/boot_modes.md:42]: sort of, I've seen it used in a couple places before[^1], but it's not widely adopted yet.

[^1]: https://0pointer.net/blog/a-re-introduction-to-mkosi-a-tool-for-generating-os-images.html#:~:text=USIs%20(Unified%20System%20Images%20which%20are%20full%20OS%20images%20packed%20in%20a%20UKI)

File: docs/boot_modes.md (hunk 2 above)
Comment (nkraetzschmar) [reply docs/boot_modes.md:52]: correct, we should be somewhere around 400-500MiB for the uncompressed rootfs which drops down to just over 200MiB for the lz4 compressed EROFS, resulting in a final UKI of under 200MiB (due to secondary compression with xz). So far away from the 4GiB limit.

Eventually the EROFS will/should be zstd compressed directly, but support for this just landed in the kernel recently(https://github.com/torvalds/linux/commit/7c35de4df1056a5a1fb4de042197b8f5b1033b61)

File: docs/boot_modes.md (hunk 3 above)
Comment (nkraetzschmar) [reply docs/boot_modes.md:22]: not specific to gardener, but to anyone consuming pre-build images. If you build your own images you won't get around reading the whole doc to decide what you want to build.

**Review by fwilhe — APPROVED** [review]
LGTM

**Pull request merged**